### ISSUE TITLE
A atst from battle 1 reacted to battle 2 and was able to trample character in battle 2

### ISSUE THREAD

**Issue by cringwell** [PlayersCommittee/gemp-swccg-public#297]
https://gemp.starwarsccg.org/gemp-swccg/game.html?replayId=advocate$ku5v76okh77kt5sj Turn 3 lightside

They missed thankfully.  But it should not be allowed to play that since it was inactive (from battling already that turn).

**Comment by Gergall** [comment]
I agree with you on those points.  1 and 2 were legal, while point 3 (playing the Trample with an inactive AT-ST) was illegal.

**Comment by cringwell** [comment]
I don't remember unfortunately.  Greg makes sense though.

**Comment by billbisco** [comment]
It looks like I stand corrected.  A very weird interaction where normally a Trample could happen anytime.

**Comment by BOTVHD** [comment]
Same issue as #554 where cards are not being seen as INACTIVE when excluded due to the rule about participating in one battle per turn.

The suggested change above (adding canBeTargetedBy(self) to the filter) is not enough to fix this - there are some more general rules/mechanics to fix.  I've added some tests showing the issues.

**Comment by BOTVHD** [comment]
As far as I can tell, this is an issue where a card not initially at the battle location moves to the battle location and is flagged as 'may not participate' by rule (namely, the "may not participate in more than one battle per turn" rule).  In this case, it is a 'react', but I think it could happen in other cases like with Path of Least Resistance.

We expect the card that moved into a second battle to have CardState.INACTIVE when checked with getCardState in this line:
```
if (!includeExcludedFromBattle && zone.isInPlay() && gameState.isDuringBattle() && isExcludedFromBattle(gameState, physicalCard))
			return CardState.INACTIVE;
```
However, the last check (isExcludedFromBattle) returns false, so instead of returning the inactive card state, code continues until eventually returning CardState.ACTIVE.

sExcludedFromBattle returns false because the 'exclude from battle' modifier was never applied to the card.  Normally, we expect that to happen via ExcludeBattleParticipantsRule (which handles excluding battle participants that aren't allowed and also adding new participants that show up after initiation).

The card that moved to the second battle is not being added to the participant list (fails to pass the canParticipateInBattleAt filter in updateParticipants?) and because there are no participants that should be excluded by rule, the ExcludeBattleParticipantsRule never applies the 'exclude from battle' modifier.
```
// Find any cards that are participating in the battle that are not allowed to be participating in the battle.
Collection<PhysicalCard> ineligibleParticipants = Filters.filterActive(game, null,
        Filters.and(Filters.in(battleState.getAllCardsParticipating()), Filters.at(battleState.getBattleLocation()), Filters.not(Filters.canParticipateInBattleAt(battleState.getBattleLocation(), battleState.getPlayerInitiatedBattle()))));
if (ineligibleParticipants.isEmpty()) {
    // Since there are no cards to exclude, just make update the participants in case any cards need to join the battle (or are no longer in the battle).
    battleState.updateParticipants(game);
    return null;
}
```

That leaves the card sort of in state of limbo during the battle.  It is not participating but it was never excluded.  That leaves it in an incorrect ACTIVE state.

It seems to me that there are a few possibilities to consider:
1. Make a change so that the card moving to battle is added as a battle participant (so the rule can apply the 'exclude from battle' modifier and remove it from the participant list)
2. Find some other place to apply the 'exclude from battle' modifier to the card
3. Change the cardState logic to catch this situation and flag the card as INACTIVE (at the battle location but not a participant?)

**Comment by BOTVHD** [comment]
For approach 3, I think this section
```
if (!includeExcludedFromBattle && zone.isInPlay() && gameState.isDuringBattle() && isExcludedFromBattle(gameState, physicalCard))
			return CardState.INACTIVE;
```

could be reworked to something like:
```
if (zone.isInPlay() && gameState.isDuringBattle() && ((isExcludedFromBattle(gameState, physicalCard) && !includeExcludedFromBattle) || mayNotParticipateByRule))
			return CardState.INACTIVE;
```

where "mayNotParticipateByRule" is a boolean thing that can catch this situation.  I'm not sure yet if it can be as simple as checking if the card passes the hasParticipatedInBattleAtOtherLocation filter (or something along those lines).

EDIT: Maybe check:
`!canParticipateInBattleAt(gameState.getBattleState().getBattleLocation(), gameState.getBattleState().getPlayerInitiatedBattle()).accepts(gameState, query(), physicalCard)`
a true result would mean the card should be inactive.  This may already incorporate checking for the ExcludedFromBattle and allow us to simplify to something like:
```
if (zone.isInPlay() && gameState.isDuringBattle() && !canParticipateInBattleAt(gameState.getBattleState().getBattleLocation(), gameState.getBattleState().getPlayerInitiatedBattle()).accepts(gameState, query(), physicalCard))
			return CardState.INACTIVE;
```

Some of these ideas won't work correctly because we don't want to make the card inactive until it arrives at the battle location (otherwise, the AT-ST wouldn't be able to react in the case above because it would incorrectly be inactive!).

Probably need to work with rules team to develop a suite of tests to check card states / exclude / participation / etc during battle.

**Comment by billbisco** [comment]
Open PR already covers this: https://github.com/PlayersCommittee/gemp-swccg-public/pull/1072

Already-battled card reacting into a second battle stays ACTIVE.
